src/contracts/SmartContract.sol

```diff
@@ -121,6 +109,22 @@ contract SmartContract {
         }
     }
 
+    function addSolution(bytes calldata solution, bytes calldata witness) public {
+        require(contractStatus == Status.Solving, "Contract is not in Solving status.");
+        require(nextUnsolvedPuzzlePart < amountOfPuzzleParts, "All puzzle parts have already been solved.");
+        require(checkSolution(solution, witness, puzzleParts[nextUnsolvedPuzzlePart].commitment), "The solution is not correct.");
+
+        uint256 upperBound = puzzleParts[nextUnsolvedPuzzlePart].upperBound;
+        require(block.timestamp <= startTime + upperBound, "Too late: the time upper bound has been exceeded.");
+
+        // Once the solution is correct, the solver should be paid
+        pay(nextUnsolvedPuzzlePart, msg.sender);
+
```

Comment:
this makes sense from how the system might work in a real-word implementation but it's misaligned with the protocol in the paper...

How about moving the checks for time and commitment to the verify function? We could call that here afterward. Similarly, payment is not automatic in the protocol but is paid out to the appropriate entity based on verification results...

Reply:
Okay, I will adjust both the SC and Mock SC to meet the specification, though it feels wrong to code this tbh 🥲

Reply:
ok,it might be worth making the protocol slightly better aligned with this